src/GraphQL/Validation/DocumentValidator.cs

```diff
@@ -81,24 +83,34 @@ public class DocumentValidator : IDocumentValidator
                 Inputs = inputs
             };
 
+            List<INodeVisitor> visitors;
+
             if (rules == null)
             {
-                rules = CoreRules;
+                // Optimizations for standard rules:
+                //   1. Known size of allocated list.
+                //   2. No LINQ related allocations.
+                //   3. No async/await related allocations since all standard rules return cached tasks from ValidateAsync.
+                var coreRules = (ReadOnlyCollection<IValidationRule>)CoreRules;
+                visitors = new List<INodeVisitor>(coreRules.Count + 1) { context.TypeInfo };
+                for (int i = 0; i < coreRules.Count; ++i)
+                {
+                    var visitor = coreRules[i].ValidateAsync(context).Result;
+                    if (visitor.ShouldRunOn(context))
+                        visitors.Add(visitor);
+                }
+            }
+            else
+            {
+                visitors = (await Task.WhenAll(rules.Select(x => x.ValidateAsync(context)))).Where(v => v.ShouldRunOn(context)).ToList();
+                visitors.Insert(0, context.TypeInfo);
```

Comment:
I don't think this "optimization" is worth the benefit, if there even is one.  Typically the rule set will inherit from `IList<IValidationRule>` and when that's the case:

1. The size of allocated list is known either way.
2. There is no need for LINQ either way.
3. There are no runtime allocations for a simple `await` on a completed task.  Since the core rules use `Task.FromResult`, there would normally be no allocations with a simple `await`.  I'm not positive but it's likely that there are no allocations for `Task.WhenAll` when all of the tasks have completed.  If this was not the case, we could optimize this like I did in the execution strategy (which would work for any rule set), but if MS did their job with `WhenAll` then it won't be any faster no matter what.

Reply:
And this code only runs once per document being validated.  Eliminating one or two allocations isn't going to have a measurable impact, even at the nanosecond level.

Reply:
This can be viewed and measured separately.